File: swift/Sources/SolanaMpp/Ed25519.swift
import Foundation
import CryptoKit

/// Ed25519 signing and verification via Apple CryptoKit.
///
/// Apple `Curve25519.Signing` implements the randomized variant
/// permitted by RFC 8032 section 5.1.6: signing the same message twice
/// with the same secret key returns two different 64-byte signatures
/// that both verify. Solana validators check only that the signature
/// verifies under the corresponding public key, so randomized output is
/// wire-compatible. The Rust and TypeScript spines use deterministic
/// `ed25519-dalek` / `@solana/keys`; both shapes are interchangeable on
/// the network because Ed25519 verification accepts any valid signature.
public enum Ed25519 {
    public static let secretKeyLength = 32
    public static let solanaKeypairLength = 64
    public static let publicKeyLength = 32
    public static let signatureLength = 64

    /// Builds a CryptoKit private key from either a 32-byte seed or the
    /// 64-byte Solana canonical keypair file format (seed `||` public key).
    public static func privateKey(from raw: Data) throws -> Curve25519.Signing.PrivateKey {
        let seed: Data
        switch raw.count {
        case secretKeyLength:
            seed = raw
        case solanaKeypairLength:
            seed = raw.prefix(secretKeyLength)
        default:
            throw MppError.signingFailure(
                "expected \(secretKeyLength) or \(solanaKeypairLength) byte secret, got \(raw.count)"
            )
        }

        do {
            return try Curve25519.Signing.PrivateKey(rawRepresentation: seed)
        } catch {
            throw MppError.signingFailure("CryptoKit rejected secret key: \(error)")
        }
    }

    public static func sign(message: Data, with privateKey: Curve25519.Signing.PrivateKey) throws -> Data {
        do {
            return try privateKey.signature(for: message)
        } catch {
            throw MppError.signingFailure("CryptoKit signature failed: \(error)")
        }
    }

    public static func verify(signature: Data, message: Data, publicKey: Data) throws -> Bool {
        guard signature.count == signatureLength else {
            throw MppError.signingFailure("signature must be \(signatureLength) bytes")
        }
        guard publicKey.count == publicKeyLength else {
            throw MppError.signingFailure("public key must be \(publicKeyLength) bytes")
        }
        do {
            let key = try Curve25519.Signing.PublicKey(rawRepresentation: publicKey)
            return key.isValidSignature(signature, for: message)
        } catch {
            throw MppError.signingFailure("CryptoKit rejected public key: \(error)")
        }
    }

    /// Returns true when the 32 bytes can be parsed as a valid Ed25519
    /// public key (i.e. lie on the curve). Used by ATA PDA derivation to
    /// reject candidate seeds whose hash lands on the curve.
    public static func isOnCurve(_ bytes: Data) -> Bool {
        guard bytes.count == publicKeyLength else { return false }
        return (try? Curve25519.Signing.PublicKey(rawRepresentation: bytes)) != nil
    }
}

File: kotlin/src/main/kotlin/com/solana/mpp/Ed25519.kt
package com.solana.mpp

import java.security.MessageDigest
import java.security.SecureRandom
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer

/**
 * Ed25519 sign / verify wrapper backed by BouncyCastle.
 *
 * BouncyCastle is used (rather than the JDK 15+ EdDSA provider) so the
 * package can import a Solana keypair from its canonical raw seed format
 * directly. The JDK provider does not expose that path uniformly across
 * supported JVMs.
 *
 * Runtime dependency: org.bouncycastle:bcprov-jdk18on:1.78.1.
 */
object Ed25519 {
    private const val SEED_LENGTH = 32
    private const val SECRET_KEY_LENGTH = 64
    private const val PUBLIC_KEY_LENGTH = 32
    private const val SIGNATURE_LENGTH = 64

    /**
     * Signs `message` deterministically against the supplied 32 byte seed.
     *
     * Returns the canonical 64 byte Ed25519 signature.
     */
    fun sign(seed: ByteArray, message: ByteArray): ByteArray {
        require(seed.size == SEED_LENGTH) {
            "Ed25519 seed must be 32 bytes (got ${seed.size})"
        }
        val key = Ed25519PrivateKeyParameters(seed, 0)
        val signer = Ed25519Signer()
        signer.init(true, key)
        signer.update(message, 0, message.size)
        return signer.generateSignature()
    }

    /** Verifies a 64 byte Ed25519 signature against a 32 byte public key. */
    fun verify(publicKey: ByteArray, message: ByteArray, signature: ByteArray): Boolean {
        require(publicKey.size == PUBLIC_KEY_LENGTH) {
            "Ed25519 public key must be 32 bytes (got ${publicKey.size})"
        }
        require(signature.size == SIGNATURE_LENGTH) {
            "Ed25519 signature must be 64 bytes (got ${signature.size})"
        }
        val key = Ed25519PublicKeyParameters(publicKey, 0)
        val verifier = Ed25519Signer()
        verifier.init(false, key)
        verifier.update(message, 0, message.size)
        return verifier.verifySignature(signature)
    }

    /** Derives the 32 byte public key for a 32 byte seed. */
    fun publicKey(seed: ByteArray): ByteArray {
        require(seed.size == SEED_LENGTH) {
            "Ed25519 seed must be 32 bytes (got ${seed.size})"
        }
        return Ed25519PrivateKeyParameters(seed, 0).generatePublicKey().encoded
    }

    /** Generates a fresh 32 byte seed using a secure random source. */
    fun generateSeed(): ByteArray {
        val random = SecureRandom()
        val seed = ByteArray(SEED_LENGTH)
        random.nextBytes(seed)
        return seed
    }

    /**
     * Normalises a raw secret-key input into a 32 byte seed.
     *
     * Solana's canonical keypair file ships 64 bytes (32 byte seed
     * concatenated with 32 byte public key). The MPP interop harness
     * passes `MPP_INTEROP_CLIENT_SECRET_KEY` as a JSON array of those 64
     * bytes. Some other tools ship the seed alone (32 bytes). This helper
     * accepts either.
     */
    fun seedFromSecretKey(secretKey: ByteArray): ByteArray = when (secretKey.size) {
        SEED_LENGTH -> secretKey.copyOf()
        SECRET_KEY_LENGTH -> secretKey.copyOfRange(0, SEED_LENGTH)
        else -> throw IllegalArgumentException(
            "Ed25519 secret key must be 32 or 64 bytes (got ${secretKey.size})",
        )
    }

    internal fun sha256(data: ByteArray): ByteArray =
        MessageDigest.getInstance("SHA-256").digest(data)
}

/** 32 byte Solana public key with base58 round-trip helpers. */
data class PublicKey(val bytes: ByteArray) {
    init {
        require(bytes.size == 32) { "Solana public key must be 32 bytes (got ${bytes.size})" }
    }

    /** Base58 encoded canonical form. */
    fun toBase58(): String = Base58.encode(bytes)

    override fun toString(): String = toBase58()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PublicKey) return false
        return bytes.contentEquals(other.bytes)
    }

    override fun hashCode(): Int = bytes.contentHashCode()

    companion object {
        /** Parses a base58 Solana address into a public key. */
        fun fromBase58(value: String): PublicKey {
            val decoded = try {
                Base58.decode(value)
            } catch (error: MppException.InvalidBase58) {
                throw MppException.InvalidPublicKey(value)
            }
            if (decoded.size != 32) {
                throw MppException.InvalidPublicKey(value)
            }
            return PublicKey(decoded)
        }
    }
}

/**
 * Solana program-derived address utilities.
 *
 * Matches `Pubkey::find_program_address` semantics used in the Rust spine
 * at `rust/src/client/charge.rs::get_associated_token_address`.
 */
object Pda {
    private val PDA_MARKER = "ProgramDerivedAddress".encodeToByteArray()
    private const val MAX_BUMP_SEED = 255

    /**
     * Finds a program-derived address by bump-iterating from 255 down to 0,
     * returning the first off-curve point (matches Solana SDK behaviour).
     */
    fun findProgramAddress(seeds: List<ByteArray>, programId: PublicKey): Pair<PublicKey, Int> {
        var bump = MAX_BUMP_SEED
        while (bump >= 0) {
            val candidateSeeds = seeds + byteArrayOf(bump.toByte())
            val candidate = try {
                createProgramAddress(candidateSeeds, programId)
            } catch (_: MppException.InvalidPublicKey) {
                bump -= 1
                continue
            }
            return candidate to bump
        }
        throw MppException.InvalidPublicKey("could not find a program-derived address")
    }

    /**
     * Hashes the seeds + program id + PDA marker to produce a candidate
     * program-derived address. Throws when the hash lands on the Ed25519
     * curve (a valid key); callers retry with a lower bump.
     */
    fun createProgramAddress(seeds: List<ByteArray>, programId: PublicKey): PublicKey {
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        for (seed in seeds) {
            if (seed.size > 32) {
                throw MppException.InvalidPublicKey("seed exceeds 32 bytes")
            }
            digest.update(seed)
        }
        digest.update(programId.bytes)
        digest.update(PDA_MARKER)
        val hash = digest.digest()
        if (isOnCurve(hash)) {
            throw MppException.InvalidPublicKey("candidate is on curve")
        }
        return PublicKey(hash)
    }

    /**
     * Returns true when the supplied 32 bytes decompress to a valid point
     * on the Ed25519 curve. Matches Solana SDK's
     * `bytes_are_curve_point` / Pubkey::is_on_curve semantics.
     */
    internal fun isOnCurve(bytes: ByteArray): Boolean = OnCurveCheck.validate(bytes)

    /** Derives the Associated Token Account for `(owner, mint, tokenProgram)`. */
    fun associatedTokenAddress(
        owner: PublicKey,
        mint: PublicKey,
        tokenProgram: PublicKey,
    ): PublicKey {
        val ataProgram = PublicKey.fromBase58(Programs.ASSOCIATED_TOKEN_PROGRAM)
        val seeds = listOf(owner.bytes, tokenProgram.bytes, mint.bytes)
        return findProgramAddress(seeds, ataProgram).first
    }
}

/**
 * On-curve check used by program-derived address derivation.
 *
 * Reuses BouncyCastle's `validatePublicKeyPartial(byte[], int)` via
 * reflection because the BC Ed25519 class is not on the BC public API
 * surface in older releases. The validation matches the reference
 * Solana SDK's `Pubkey::is_on_curve` semantics: a 32 byte value is on
 * the curve when it decompresses to a valid (x, y) point on Ed25519.
 */
internal object OnCurveCheck {
    private val method = run {
        val clazz = Class.forName("org.bouncycastle.math.ec.rfc8032.Ed25519")
        val m = clazz.getDeclaredMethod(
            "validatePublicKeyPartial",
            ByteArray::class.java,
            Int::class.javaPrimitiveType,
        )
        m.isAccessible = true
        m
    }

    fun validate(bytes: ByteArray): Boolean {
        if (bytes.size != 32) return false
        val result = method.invoke(null, bytes, 0)
        return result as Boolean
    }
}
